core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpooledQueryDataProducer.java

```diff
@@ -44,7 +44,7 @@ public class SpooledQueryDataProducer
         implements QueryDataProducer
 {
     private final QueryDataEncoder.Factory encoderFactory;
-    private final AtomicBoolean metadataWritten = new AtomicBoolean(false);
+    private final AtomicReference<QueryDataEncoder> encoder = new AtomicReference<>();
```

Comment:
make it normal mutable field if you do not need to guard from concurrency. Use of AR is misleading.